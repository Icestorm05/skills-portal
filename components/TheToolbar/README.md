# The Toolbar

## Functionality

The toolbar displays the skills portal logo. When logged in, it also displays: the avatar of the user logged in, a logout button, and a slot for a tabs component. This component should only be used once within the lifecycle of this application.

### Props

`<the-toolbar>` supports the following custom component attributes:

| attribute | type | description
| --- | --- | ---
| `color` | String *optional* | The color of the toolbar. Defaults to `primary`. Accepts any of: `primary`, `accent`, `secondary`, `info`, `warning`, `error` or `success`.
| `isLoggedIn` | Boolean *optional* | Whether to display the logged in restricted components or not. Defaults to `false`.

### Events

`<the-toolbar>` exposes the following events:

| event | description
| --- | ---
| `avatar-click` | Fired when the avatar is clicked.
| `help-click` | Fired when the help button is clicked.
| `logo-click` | Fired when the logo is clicked.
| `logout-click` | Fired when the logout button is clicked.

### Slots

`<the-toolbar>` exposes the following slots:

| slot | description
| --- | ---
`tabs` | The tabs to display in the extension of the toolbar.

### Example

```html
<the-toolbar color="secondary"
             @avatar-click="home"
             @help-click="help"
             @logo-click="home"
             @logout-click="logout">
    <nav slot="tabs">
        <a href="/test">Test Page</a>
    </nav>
</the-toolbar>
```

```ts
import Vue from 'vue';
import Component from 'nuxt-class-component';
import TheToolbar from '~/components/TheToolbar/TheToolbar.vue';
@Component({
    components: {
        TheToolbar
    }
})
export default class GenericComponent extends Vue {
    private help() {
        this.$router.push('/faq');
    }
    private home() {
        this.$router.push('/');
    }
    private logout() {
        this.$router.push('/login');
    }
}
```