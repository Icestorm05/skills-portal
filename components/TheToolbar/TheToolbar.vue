<template lang="pug">
  v-toolbar(app
            :color="color"
            fixed
            :extension-height="72")#TheToolbar
    img(alt="Skills Portal Logo"
        height="65%"
        src="~/assets/img/skills_logo.png"
        width="auto"
        @click="$emit('logo-click')")#TheToolbarSkillsLogo.ml-5
    v-spacer
    v-menu(offset-y
           v-show="isLoggedIn")#TheToolbarAvatarMenu
        v-avatar(slot="activator"
                 @click="$emit('avatar-click')")#TheToolbarAvatar.mr-3
            avatar-image(v-if="isLoggedIn" :image-name="avatarImageName" :renderAsBase64="true")
        v-list
            v-list-tile(@click="$emit('profile-click')"
                        nuxt)#TheToolbarProfile View Profile
            v-list-tile(@click="$emit('password-click')"
                        nuxt)#TheToolbarPassword Change Password
    v-btn(v-show="isLoggedIn"
          color="white"
          flat
          icon
          aria-label="Help"
          @click="$emit('help-click')")#TheToolbarHelpButton
        v-icon help_outline
    v-btn(v-show="isLoggedIn"
          color="white"
          flat
          icon
          aria-label="Logout"
          @click="$emit('logout-click')")#TheToolbarLogoutButton
        v-icon exit_to_app
    slot(slot="extension" name="tabs")
</template>

<script lang="ts">
import Component from "nuxt-class-component";
import Vue from "vue";
import { Prop } from "vue-property-decorator";
import { Getter } from "vuex-class";

import VAvatar from "vuetify/es5/components/VAvatar";
import VBtn from "vuetify/es5/components/VBtn";
import { VSpacer } from "vuetify/es5/components/VGrid";
import VIcon from "vuetify/es5/components/VIcon";
import VList, { VListTile } from "vuetify/es5/components/VList";
import VMenu from "vuetify/es5/components/VMenu";
import VToolbar, { VToolbarTitle } from "vuetify/es5/components/VToolbar";
import AvatarImage from "~/components/AvatarImage/AvatarImage.vue";

@Component({
  components: {
    AvatarImage,
    VAvatar,
    VBtn,
    VSpacer,
    VIcon,
    VList,
    VListTile,
    VMenu,
    VToolbar,
    VToolbarTitle
  }
})
export default class TheToolbar extends Vue {
  @Prop({ type: String, default: "" })
  private avatarImageName: string;
  @Prop({ type: String, default: "primary" })
  private color: string;
  @Prop({ type: Boolean, default: false })
  private isLoggedIn: boolean;
}
</script>

<style lang="stylus" scoped>
#TheToolbar {
    #TheToolbarSkillsLogo {
        cursor: pointer;
    }

    #TheToolbarAvatar {
        cursor: pointer;
    }
}
</style>