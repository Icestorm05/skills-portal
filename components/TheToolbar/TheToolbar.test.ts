import TheToolbar from "./TheToolbar.vue";
import Vuetify from "vuetify/es5/components/Vuetify";
import VToolbar from "vuetify/es5/components/VToolbar";
import { createLocalVue, mount } from "@vue/test-utils";

declare const expect: jest.Expect;

const localVue = createLocalVue();
localVue.use(Vuetify, {
  components: {
    VToolbar
  }
});

// Add a wrapping <div data-app="true"></div> to silence Vuetify warnings.
const app = document.createElement("div");
app.setAttribute("data-app", "true");
document.body.appendChild(app);

describe("The Toolbar", () => {
  test("is a Vue instance", () => {
    const wrapper = mount(TheToolbar, { localVue });
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test("snapshot has not changed", () => {
    const wrapper = mount(TheToolbar, { localVue });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("color prop changes color of toolbar", () => {
    const wrapper = mount(TheToolbar, {
      localVue,
      propsData: { color: "warning" }
    });
    expect(wrapper.find("#TheToolbar").classes()).toContain("warning");
  });

  test("clicking on the skills logo emits the logo-click event", () => {
    const wrapper = mount(TheToolbar, { localVue });
    wrapper.find("#TheToolbar #TheToolbarSkillsLogo").trigger("click");
    expect(wrapper.emitted("logo-click")).toBeTruthy();
  });

  test("avatar is hidden by default", () => {
    const wrapper = mount(TheToolbar, { localVue });
    expect(
      wrapper.find("#TheToolbar #TheToolbarAvatar").isVisible()
    ).toBeFalsy();
  });

  test("avatar is displayed with isLoggedIn prop", () => {
    const wrapper = mount(TheToolbar, {
      localVue,
      propsData: { isLoggedIn: true }
    });
    expect(
      wrapper.find("#TheToolbar #TheToolbarAvatar").isVisible()
    ).toBeTruthy();
  });

  test("clicking on view profile emits the profile-click event", () => {
    const wrapper = mount(TheToolbar, {
      localVue,
      propsData: { isLoggedIn: true }
    });
    wrapper.find("#TheToolbarProfile").trigger("click");
    expect(wrapper.emitted("profile-click")).toBeTruthy();
  });

  test("clicking on change password emits the password-click event", () => {
    const wrapper = mount(TheToolbar, {
      localVue,
      propsData: { isLoggedIn: true }
    });
    wrapper.find("#TheToolbarPassword").trigger("click");
    expect(wrapper.emitted("password-click")).toBeTruthy();
  });

  test("help button is hidden by default", () => {
    const wrapper = mount(TheToolbar, { localVue });
    expect(
      wrapper.find("#TheToolbar #TheToolbarHelpButton").isVisible()
    ).toBeFalsy();
  });

  test("help buton os displayed with isLoggedIn prop", () => {
    const wrapper = mount(TheToolbar, {
      localVue,
      propsData: { isLoggedIn: true }
    });
    expect(
      wrapper.find("#TheToolbar #TheToolbarHelpButton").isVisible()
    ).toBeTruthy();
  });

  test("clicking on help button emits the help-click event", () => {
    const wrapper = mount(TheToolbar, {
      localVue,
      propsData: { isLoggedIn: true }
    });
    wrapper.find("#TheToolbar #TheToolbarHelpButton").trigger("click");
    expect(wrapper.emitted("help-click")).toBeTruthy();
  });

  test("logout button is hidden by default", () => {
    const wrapper = mount(TheToolbar, { localVue });
    expect(
      wrapper.find("#TheToolbar #TheToolbarLogoutButton").isVisible()
    ).toBeFalsy();
  });

  test("logout button is displayed with isLoggedIn prop", () => {
    const wrapper = mount(TheToolbar, {
      localVue,
      propsData: { isLoggedIn: true }
    });
    expect(
      wrapper.find("#TheToolbar #TheToolbarLogoutButton").isVisible()
    ).toBeTruthy();
  });

  test("clicking on logout button emits the logout-click event", () => {
    const wrapper = mount(TheToolbar, {
      localVue,
      propsData: { isLoggedIn: true }
    });
    wrapper.find("#TheToolbar #TheToolbarLogoutButton").trigger("click");
    expect(wrapper.emitted("logout-click")).toBeTruthy();
  });

  test("tabs slot displays content", () => {
    const wrapper = mount(TheToolbar, {
      localVue,
      slots: {
        tabs: `<nav id="TestNav">
                    <a>Test Item</a>
                   </nav>`
      }
    });
    expect(wrapper.find("#TheToolbar #TestNav").exists()).toBeTruthy();
  });

  test("help button is hidden by default", () => {
    const wrapper = mount(TheToolbar, { localVue });
    expect(
      wrapper.find("#TheToolbar #TheToolbarHelpButton").isVisible()
    ).toBeFalsy();
  });

  test("help button is displayed with isLoggedIn prop", () => {
    const wrapper = mount(TheToolbar, {
      localVue,
      propsData: { isLoggedIn: true }
    });
    expect(
      wrapper.find("#TheToolbar #TheToolbarHelpButton").isVisible()
    ).toBeTruthy();
  });

  test("clicking on help button emits the help-click event", () => {
    const wrapper = mount(TheToolbar, {
      localVue,
      propsData: { isLoggedIn: true }
    });
    wrapper.find("#TheToolbar #TheToolbarHelpButton").trigger("click");
    expect(wrapper.emitted("help-click")).toBeTruthy();
  });
});
