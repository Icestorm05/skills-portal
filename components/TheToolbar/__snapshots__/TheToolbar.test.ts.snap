// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Toolbar snapshot has not changed 1`] = `
<nav class="toolbar toolbar--fixed primary" style="margin-top: 0px; padding-right: 0px; padding-left: 0px;" id="TheToolbar">
    <div class="toolbar__content" style="height: 56px;">
        <img alt="Skills Portal Logo" height="65%" src="~/assets/img/skills_logo.png" width="auto" id="TheToolbarSkillsLogo" class="ml-5">
        <div class="spacer"></div>
        <div class="menu" style="display: none;" id="TheToolbarAvatarMenu">
            <div class="menu__activator">
                <div id="TheToolbarAvatar" class="avatar mr-3" style="height: 48px; width: 48px;">
                    <!---->
                </div>
            </div>
        </div>
        <button type="button" class="btn btn--flat btn--icon white--text" aria-label="Help" id="TheToolbarHelpButton" style="display: none;">
            <div class="btn__content">
                <i aria-hidden="true" class="icon material-icons">help_outline</i>
            </div>
        </button>
        <button type="button" class="btn btn--flat btn--icon white--text" aria-label="Logout" id="TheToolbarLogoutButton" style="display: none;">
            <div class="btn__content">
                <i aria-hidden="true" class="icon material-icons">exit_to_app</i>
            </div>
        </button>
    </div>
</nav>
`;
